threedi_schema/application/upgrade_utils.py

```diff
@@ -0,0 +1,78 @@
+import logging
+from typing import Callable, TYPE_CHECKING
+
+from alembic.config import Config
+from alembic.script import ScriptDirectory
+
+if TYPE_CHECKING:
+    from .schema import ModelSchema
+else:
+    ModelSchema = None
+
+
+class ProgressHandler(logging.Handler):
+    def __init__(self, progress_func, total_steps):
+        super().__init__()
+        self.progress_func = progress_func
+        self.total_steps = total_steps
+        self.current_step = 0
+
+    def emit(self, record):
+        msg = record.getMessage()
+        if msg.startswith("Running upgrade"):
+            self.progress_func(100 * self.current_step / self.total_steps)
+            self.current_step += 1
+
+
+def get_upgrade_steps_count(
+    config: Config, current_revision: int, target_revision: str = "head"
+) -> int:
+    """
+    Count number of upgrade steps for a schematisation upgrade.
+
+    Args:
+        config: Config parameter containing the configuration information
+        current_revision: current revision as integer
+        target_revision: target revision as zero-padded 4 digit string or "head"
+    """
+    if target_revision != "head":
+        try:
+            int(target_revision)
+        except TypeError:
+            # this should lead to issues in the upgrade pipeline, lets not take over that error handling here
+            return 0
+    offset = 0
+    if current_revision < 200:
+        current_revision = 200
+        offset = 1
```

Comment:
what is this doing?

Reply:
`walk_revisions` cannot handle revision numbers before 200 because there is no revision < 200 defined. The upgrade from current to 200 is one step. So by starting counting from 200 and adding 1 step we can count by starting from 200. 

See some slight changes and added comments in 2e94b367410da17e13ceab1331f737231c4a0edc